### PR TITLE
routing+routerrpc: add multiple outgoing channel restriction

### PR DESCRIPTION
**Pull request by joostjager**
This PR adds the ability to restrict payment pathfinding to a set of outgoing channels. Prior to this, it was only possible to restrict to a single channel.

A use case of channel (set) selection is to not just make the payment, but also obtain the (right) inbound at the same time. This is also a building block that allows loop out channel set selection.

### CODE REVIEW

File: routing/pathfind.go [hunk 1]
```diff
@@ -329,7 +329,7 @@ type PathFindingConfig struct {
 // getOutgoingBalance returns the maximum available balance in any of the
 // channels of the given node. The second return parameters is the total
 // available balance.
-func getOutgoingBalance(node route.Vertex, outgoingChan *uint64,
+func getOutgoingBalance(node route.Vertex, outgoingChan map[uint64]struct{},
```
Comment (Roasbeef) [inline routing/pathfind.go:332]: `outgoingChan` -> `outgoingChans`

File: routing/pathfind_test.go [hunk 2]
```diff
@@ -2306,6 +2309,17 @@ func TestRestrictOutgoingChannel(t *testing.T) {
 		t.Fatalf("expected route to pass through channel 2, "+
 			"but channel %v was selected instead", route.Hops[0].ChannelID)
 	}
+
+	// If a direct channel to target is allowed as well, that channel is
+	// expected to be selected because the routing fees are zero.
+	ctx.restrictParams.OutgoingChannelIDs = []uint64{2, 6}
```
Comment (Roasbeef) [inline routing/pathfind_test.go:2315]: Where do these chan IDs come from? The order they're added to the graph above? If so would be nice to move them into a constant, or perhaps add a method to `[]*testChannel` which lets you look them up easily.

File: routing/pathfind_test.go [hunk 3]
```diff
@@ -2306,6 +2309,17 @@ func TestRestrictOutgoingChannel(t *testing.T) {
...
+	path, err = ctx.findPath(target, paymentAmt)
+	if err != nil {
+		t.Fatalf("unable to find path: %v", err)
+	}
+	if path[0].ChannelID != 6 {
+		t.Fatalf("expected routt to pass through channel 6")
```
Comment (Roasbeef) [inline routing/pathfind_test.go:2321]: routt -> route

File: routing/pathfind.go (hunk 1 above)
Comment (joostjager) [reply routing/pathfind.go:332]: fixed

File: routing/pathfind_test.go (hunk 3 above)
Comment (joostjager) [reply routing/pathfind_test.go:2321]: fixed

File: routing/pathfind_test.go (hunk 2 above)
Comment (joostjager) [reply routing/pathfind_test.go]: constants added

File: lnrpc/routerrpc/router_backend.go [hunk 4]
```diff
@@ -522,9 +522,14 @@ func (r *RouterBackend) extractIntentFromSendRequest(
...
 	payIntent := &routing.LightningPayment{}
 
-	// Pass along an outgoing channel restriction if specified.
+	// Pass along restrictions on the outgoing channels that may be used.
+	payIntent.OutgoingChannelIDs = rpcPayReq.OutgoingChanIds
+
+	// Add the deprecated single outgoing channel restriction if present.
 	if rpcPayReq.OutgoingChanId != 0 {
-		payIntent.OutgoingChannelID = &rpcPayReq.OutgoingChanId
+		payIntent.OutgoingChannelIDs = append(
```
Comment (carlaKC) [inline lnrpc/routerrpc/router_backend.go:530]: Could also fail if both are set? Take the deprecation of the old field seriously if the caller is using the new one?

Comment (joostjager) [reply lnrpc/routerrpc/router_backend.go:535]: Yes, I want to be serious about it. Fixed

**Comment by joostjager** [comment]
Tested channel set restriction with loop out on a dev branch, works.

**Pull request merged**